File: commonfamily/src/main/java/com/koenig/communication/messages/AskForUpdatesMessage.java
package com.koenig.communication.messages;

import com.koenig.commonModel.Byteable;
import com.koenig.commonModel.Component;
import com.koenig.commonModel.ItemType;

import org.joda.time.DateTime;

import java.nio.ByteBuffer;

/**
 * Created by Thomas on 30.11.2017.
 */

public class AskForUpdatesMessage extends FamilyMessage {
    public static final String NAME = "AskForUpdatesMessage";
    DateTime lastSyncDate;
    ItemType updateType;

    public AskForUpdatesMessage(Component component, DateTime lastSyncDate, ItemType updateType) {
        super(component);
        this.lastSyncDate = lastSyncDate;
        this.updateType = updateType;
    }

    public AskForUpdatesMessage(int version, Component component, String fromId, String toId, ByteBuffer buffer) {
        super(component);
        this.version = version;
        this.fromId = fromId;
        this.toId = toId;
        this.lastSyncDate = new DateTime(buffer.getLong());
        this.updateType = Byteable.Companion.byteToEnum(buffer, ItemType.class);
    }

    public static AskForUpdatesMessage askForExpenses(DateTime lastSyncDate) {
        return new AskForUpdatesMessage(Component.FINANCE, lastSyncDate, ItemType.EXPENSES);
    }

    public static AskForUpdatesMessage askForStandingOrders(DateTime lastSyncDate) {
        return new AskForUpdatesMessage(Component.FINANCE, lastSyncDate, ItemType.STANDING_ORDER);
    }

    public static AskForUpdatesMessage askForCategorys(DateTime lastSyncDate) {
        return new AskForUpdatesMessage(Component.FINANCE, lastSyncDate, ItemType.CATEGORY);
    }

    public ItemType getUpdateType() {
        return updateType;
    }

    public String getName() {
        return NAME;
    }

    public DateTime getLastSyncDate() {
        return lastSyncDate;
    }


    @Override
    protected int getContentLength() {
        return Byteable.Companion.getDateLength() + Byteable.Companion.getEnumLength(updateType);
    }

    @Override
    protected void writeContent(ByteBuffer buffer) {
        Byteable.Companion.writeDateTime(lastSyncDate, buffer);
        Byteable.Companion.writeEnum(updateType, buffer);
    }

    @Override
    public String toString() {
        return "AskForUpdatesMessage{" +
                "lastSyncDate=" + lastSyncDate +
                ", updateType='" + updateType + '\'' +
                '}';
    }
}

File: commonfamily/src/main/java/com/koenig/communication/messages/AskForUpdatesMessage.kt
package com.koenig.communication.messages

import com.koenig.commonModel.Byteable
import com.koenig.commonModel.Component
import com.koenig.commonModel.ItemType

import org.joda.time.DateTime

import java.nio.ByteBuffer

/**
 * Created by Thomas on 30.11.2017.
 */

class AskForUpdatesMessage : FamilyMessage {
    var lastSyncDate: DateTime
        internal set
    var updateType: ItemType
        internal set

    override val name: String
        get() = NAME


    override val contentLength: Int
        get() = Byteable.dateLength + Byteable.getEnumLength(updateType)

    constructor(component: Component, lastSyncDate: DateTime, updateType: ItemType) : super(component) {
        this.lastSyncDate = lastSyncDate
        this.updateType = updateType
    }

    constructor(version: Int, component: Component, fromId: String, toId: String, messageId: String, buffer: ByteBuffer) : super(component, messageId) {
        this.version = version
        this.fromId = fromId
        this.toId = toId
        this.lastSyncDate = DateTime(buffer.long)
        this.updateType = Byteable.byteToEnum(buffer, ItemType::class.java)
    }

    override fun writeContent(buffer: ByteBuffer) {
        Byteable.writeDateTime(lastSyncDate, buffer)
        Byteable.writeEnum(updateType, buffer)
    }

    override fun toString(): String {
        return "AskForUpdatesMessage{" +
                "lastSyncDate=" + lastSyncDate +
                ", updateType='" + updateType + '\''.toString() +
                '}'.toString()
    }

    companion object {
        val NAME = "AskForUpdatesMessage"

        fun askForExpenses(lastSyncDate: DateTime): AskForUpdatesMessage {
            return AskForUpdatesMessage(Component.FINANCE, lastSyncDate, ItemType.EXPENSES)
        }

        fun askForStandingOrders(lastSyncDate: DateTime): AskForUpdatesMessage {
            return AskForUpdatesMessage(Component.FINANCE, lastSyncDate, ItemType.STANDING_ORDER)
        }

        fun askForCategorys(lastSyncDate: DateTime): AskForUpdatesMessage {
            return AskForUpdatesMessage(Component.FINANCE, lastSyncDate, ItemType.CATEGORY)
        }
    }
}
